text/0000-unsafe-attributes.md

```diff
@@ -0,0 +1,175 @@
+  feels much less systematic and much more ad-hoc.
+- **`deny(unsafe_code)`.** We already
+  [started the process](https://github.com/rust-lang/rust/issues/82499) of
+  rejecting these attributes when `deny(unsafe_code)` is used. We could say that
+  is enough. However the RFC authors thinks that is insufficient, since only few
+  crates use that lint, and since it is the wrong default for Rust (users have
+  to opt-in to a soundness-critical diagnostic -- that's totally against the
+  "safety by default" goal of Rust). This RFC says that yes, `deny(unsafe_code)`
+  should deny those attributes, but we should go further and require an explicit
+  `unsafe(...)` attribute block for them to be used at all.
+- **Item-level unsafe blocks.** We could find some way to have 'unsafe blocks'
+  around entire functions or modules. However, those would go against the usual
+  goal of keeping `unsafe` blocks small. Big `unsafe` blocks risk accidentally
+  calling an unsafe operation in there without even realizing it.
+- **Other syntax.** Obviously we could pick a different syntax for the same
+  concept, but this seems like the most natural marriage of the idea of unsafe
+  blocks from regular code, and the existing attributes syntax.
+
+# Prior art
+[prior-art]: #prior-art
+
+We have `unsafe` blocks; this is basically the same thing for the "attributes
+DSL".
+
+In the attribute DSL, we already have a "nesting" construct: `cfg_attr`. That
+allows terms like
+`#[cfg_attr(debug_assertions, deny(unsafe_code), allow(unused))]`, so there is
+precedent for having a list of attributes inside a single attribute.
+
+I don't know of other languages that would distinguish safe and unsafe
+attributes.
+
+# Unresolved questions
+[unresolved-questions]: #unresolved-questions
+
+- **Different lint staging.** The lint on using existing unsafe attributes like
+  `no_mangle` outside `unsafe(...)` could be staged in various ways: it could be
+  warn-by-default to start or we wait a while before to do that, it could be
+  edition-dependent, it might eventually be deny-by-default or even a hard error
+  on some editions -- there are lots of details here, which can be determined
+  later during the process.
+
+# Future possibilities
+[future-possibilities]: #future-possibilities
+
+- **Unsafe attribute proc macros.** We could imagine something like
+    ```
+    #[proc_macro_attribute(require_unsafe)]
+    fn spoopy(args: TokenStream, input: TokenStream) -> TokenStream {…}
+    ```
+  to declare that an attribute proc macro is unsafe to use, and must only
+  occur as an unsafe macro. Such an unsafe-to-use attribute proc macro must
+  declare in a comment what its safety requirements are. (This is the `unsafe`
+  from `unsafe fn`, whereas the rest of the RFC is using the `unsafe` from
+  `unsafe { ... }`.)
+- **Unsafe derive.** We could use `#[unsafe(derive(Trait))]` to derive an
+  `unsafe impl` where the deriving macro itself cannot check all required safety
+  conditions.
```

Comment:
I don't want to bikeshed something that doesn't need to be in this RFC. But I do think this is the wrong syntax. unsafe used like that should be "unsafe to derive", not "derives an unsafe trait".

Reply:
Indeed, the intention is that this is an example for "unsafe to derive".